src/lib.rs

```diff
@@ -510,40 +330,47 @@ where
         })
     }
 
-    /// Create a builder for a batch publish construction.
-    pub fn build_publish(&self) -> HedwigPublishBuilder<T, P> {
-        HedwigPublishBuilder {
+    /// Create a batch of messages to publish
+    ///
+    /// This allows to transparently retry failed messages and send them in batches larger than
+    /// one, leading to potential throughput gains.
+    pub fn build_batch(&self) -> PublishBatch<T, P> {
+        PublishBatch {
             hedwig: self,
             messages: Vec::new(),
         }
     }
 
-    /// A convenience shortcut to publish a single message.
+    /// Publish a single message
     ///
-    /// Equivalent to `hedwig.build_publish().message(msg)?.publish()`.
-    pub fn publish<D>(&self, msg: Message<D, T>) -> Result<P::MessageIds, PublishError>
+    /// Note, that unlike the batch builder, this does not allow recovering failed-to-publish
+    /// messages.
+    pub async fn publish<D>(&self, msg: Message<D, T>) -> Result<P::MessageId, Error>
     where
-        D: Serialize,
+        D: serde::Serialize,
         T: Copy + Into<&'static str>,
     {
-        let mut builder = self.build_publish();
+        let mut builder = self.build_batch();
         builder.message(msg)?;
-        builder.publish()
+        builder.publish_one().await
     }
 }
 
-/// A builder for batch publish.
+/// A builder for publishing in batches
+///
+/// Among other things this structure also enables transparent retrying of failed-to-publish
+/// messages.
 #[allow(missing_debug_implementations)]
-pub struct HedwigPublishBuilder<'hedwig, T, P> {
+pub struct PublishBatch<'hedwig, T, P> {
     hedwig: &'hedwig Hedwig<T, P>,
     messages: Vec<(&'static str, ValidatedMessage)>,
 }
 
-impl<'hedwig, T, P> HedwigPublishBuilder<'hedwig, T, P> {
-    /// Add a message to be published in a batch.
-    pub fn message<D>(&mut self, msg: Message<D, T>) -> Result<&mut Self, PublishError>
+impl<'hedwig, T, P> PublishBatch<'hedwig, T, P> {
+    /// Add a message to be published in a batch
+    pub fn message<D>(&mut self, msg: Message<D, T>) -> Result<&mut Self, Error>
```

Comment:
not entirely sure I understand how async and borrowing rules work but would this method be prevented automatically _while_ a publish is happening?

Reply:
Yes.